cpp/src/arrow/compute/kernels/vector_swizzle.cc

```diff
@@ -0,0 +1,415 @@
+                             NullFunc&& null_func) {
+    return VisitArraySpanInline<IndexType>(span, std::forward<ValidFunc>(valid_func),
+                                           std::forward<NullFunc>(null_func));
+  }
+
+  static Status Exec(KernelContext* ctx, const ExecSpan& span, ExecResult* result) {
+    DCHECK_EQ(span.num_values(), 1);
+    DCHECK(span[0].is_array());
+    const auto& indices = span[0].array;
+    ARROW_ASSIGN_OR_RAISE(
+        result->value, InversePermutationImpl<ThisType>::Exec(
+                           ctx, indices, indices.length, indices.type->GetSharedPtr()));
+    return Status::OK();
+  }
+};
+
+template <typename Ignored, typename Type>
+struct InversePermutationChunked {
+  using ThisType = InversePermutationChunked<Ignored, Type>;
+  using IndexType = Type;
+  using ShapeType = std::shared_ptr<ChunkedArray>;
+
+  template <typename ValidFunc, typename NullFunc>
+  static Status VisitIndices(const std::shared_ptr<ChunkedArray>& chunked_array,
+                             ValidFunc&& valid_func, NullFunc&& null_func) {
+    for (const auto& chunk : chunked_array->chunks()) {
+      ArraySpan span(*chunk->data());
+      RETURN_NOT_OK(VisitArraySpanInline<IndexType>(
+          span, std::forward<ValidFunc>(valid_func), std::forward<NullFunc>(null_func)));
+    }
+    return Status::OK();
+  }
+
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* result) {
+    DCHECK_EQ(batch.num_values(), 1);
+    DCHECK(batch[0].is_chunked_array());
+    const auto& indices = batch[0].chunked_array();
+    ARROW_ASSIGN_OR_RAISE(auto inverse_permutation,
+                          InversePermutationImpl<ThisType>::Exec(
+                              ctx, indices, indices->length(), indices->type()));
+    *result =
+        Datum(std::make_shared<ChunkedArray>(MakeArray(std::move(inverse_permutation))));
+    return Status::OK();
+  }
+};
+
+void RegisterVectorInversePermutation(FunctionRegistry* registry) {
+  auto function = std::make_shared<VectorFunction>("inverse_permutation", Arity::Unary(),
+                                                   inverse_permutation_doc,
+                                                   GetDefaultInversePermutationOptions());
+
+  auto add_kernel = [&function](Type::type type_id) {
+    VectorKernel kernel;
+    kernel.signature =
+        KernelSignature::Make({InputType(match::SameTypeId(type_id))},
+                              OutputType(ResolveInversePermutationOutputType));
+    kernel.init = InversePermutationState::Init;
+    kernel.exec = GenerateInteger<InversePermutation, void, ArrayKernelExec>(type_id);
+    kernel.exec_chunked =
+        GenerateInteger<InversePermutationChunked, void, VectorKernel::ChunkedExec>(
+            type_id);
+    kernel.can_execute_chunkwise = false;
+    kernel.output_chunked = false;
+    DCHECK_OK(function->AddKernel(std::move(kernel)));
+  };
+  for (const auto& t : IntTypes()) {
```

Comment:
Is it worth defining this function on both signed and unsigned types? We should be mindful of code generation complexity, and this function instantiates a cartesian product of input type and desired output type. Limiting ourselves to signed values (for example) would divide code generation by 4.

Reply:
I think signed integers should suffice. I've reduced the input/output types to signed only.